File: iosApp/iosApp/Modules/DegenCult ✅/CultProposals ✅/CultProposalsWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - CultProposalsWireframeFactory

protocol CultProposalsWireframeFactory {

    func make(_ parent: UIViewController?) -> CultProposalsWireframe
}

// MARK: - DefaultCultProposalsWireframeFactory

final class DefaultCultProposalsWireframeFactory {

    private let cultProposalWireframeFactory: CultProposalWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let cultService: CultService
    private let walletService: WalletService
    private let networksService: NetworksService

    init(
        cultProposalWireframeFactory: CultProposalWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        cultService: CultService,
        walletService: WalletService,
        networksService: NetworksService
    ) {
        self.cultProposalWireframeFactory = cultProposalWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.cultService = cultService
        self.walletService = walletService
        self.networksService = networksService
    }
}

extension DefaultCultProposalsWireframeFactory: CultProposalsWireframeFactory {

    func make(_ parent: UIViewController?) -> CultProposalsWireframe {
        DefaultCultProposalsWireframe(
            parent,
            cultProposalWireframeFactory: cultProposalWireframeFactory,
            confirmationWireframeFactory: confirmationWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            currencySwapWireframeFactory: currencySwapWireframeFactory,
            cultService: cultService,
            walletService: walletService,
            networksService: networksService
        )
    }
}

// MARK: - Assembler

final class CultProposalsWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> CultProposalsWireframeFactory in
            DefaultCultProposalsWireframeFactory(
                cultProposalWireframeFactory: resolver.resolve(),
                confirmationWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                currencySwapWireframeFactory: resolver.resolve(),
                cultService: resolver.resolve(),
                walletService: resolver.resolve(),
                networksService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/cultproposals/CultProposalsWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.cultproposals

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.cultproposal.CultProposalWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.degenCultProposals.CultProposalsWireframe
import com.sonsofcrypto.web3walletcore.services.cult.CultService
import smartadapter.internal.extension.name

interface CultProposalsWireframeFactory {
    fun make(parent: Fragment?): CultProposalsWireframe
}

class DefaultCultProposalsWireframeFactory(
    private val cultService: CultService,
    private val walletService: WalletService,
    private val networksService: NetworksService,
    private val cultProposalWireframeFactory: CultProposalWireframeFactory
): CultProposalsWireframeFactory {

    override fun make(parent: Fragment?): CultProposalsWireframe = DefaultCultProposalsWireframe(
        parent?.let { WeakRef(it) },
        cultService,
        walletService,
        networksService,
        cultProposalWireframeFactory,
    )
}

class CultProposalsWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CultProposalsWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultCultProposalsWireframeFactory(
                it.resolve(CultService::class.name),
                it.resolve(WalletService::class.name),
                it.resolve(NetworksService::class.name),
                it.resolve(CultProposalWireframeFactory::class.name),
            )
        }
    }
}